File: 707_software_patterns/assn/01/src/swift/Patterns/Patterns/XMLValidator.swift
//
//  XMLValidator.swift
//  Patterns
//
//  Created by Jon Lindberg on 1/1/19.
//  Copyright © 2019 Jon Lindberg. All rights reserved.
//

import Foundation

public class ValidChildren
{
	public init(thisElement : String)
	{
		thisElementImpl	= thisElement
	}
	
	internal let thisElementImpl	: String
	internal var validChildren	= [String]()
	internal var childIsAttribute	= [Bool]()
	internal var canHaveTextImpl	= false
	
	public var thisElement	: String	{ get { return thisElementImpl } }
	public var canHaveText : Bool		{ get { return canHaveTextImpl } set { canHaveTextImpl = newValue } }

	public func addValidChild(child : String, isAttribute : Bool)
	{
		if childIsValid(child: child, isAttribute: isAttribute)
		{
			return
		}
		
		validChildren.append(child)
		childIsAttribute.append(isAttribute)
	}
	public func childIsValid(child : String, isAttribute : Bool) -> Bool
	{
		for i in 0..<validChildren.count
		{
			if childIsAttribute[i] == isAttribute && validChildren[i] == child
			{
				return true
			}
		}
		
		return false
	}
}

public class XMLValidator
{
	init() {}
	
	internal var schema	= [ValidChildren]()
	
	public func addSchemaElement(element : String)				-> ValidChildren
	{
		let schemaElementIterator	= findSchemaElement(element: element)
		
		if schemaElementIterator < schema.count
		{
			schema.remove(at: schemaElementIterator)
		}
		
		let schemaElement	= ValidChildren(thisElement: element)
		schema.append(schemaElement)
		return schemaElement
	}
	public func findSchemaElement(element : String)				-> Int
	{
		for i in 0..<schema.count
		{
			if (schema[i].thisElement.count == 0 && element.count == 0) ||
			  (schema[i].thisElement.count > 0 && schema[i].thisElement == element)
			{
				return i
			}
		}
		
		return schema.count
	}
	public func canRootElement(newElement : String)				-> Bool
	{
		return canAddElement(element: nil, newElement: newElement)
	}
	public func canAddElement(element : Element?, newElement : String)	-> Bool
	{
		let schemaElement	= findSchemaElement(element: element === nil ? "" : element!.tagName)
		
		return
		  schemaElement >= schema.count ? true : schema[schemaElement].childIsValid(child: newElement, isAttribute: false)
	}
	public func canAddText(element : Element)				-> Bool
	{
		let schemaElement	= findSchemaElement(element: element.tagName)
		
		return schemaElement >= schema.count ? true : schema[schemaElement].canHaveText
	}
	public func canAddAttribute(element : Element, newAttribute : String)	-> Bool
	{
		let schemaElement	= findSchemaElement(element: element.tagName)
		
		return schemaElement >= schema.count ?
		  true : schema[schemaElement].childIsValid(child: newAttribute, isAttribute: true)
	}
}

File: 707_software_patterns/assn/03/prog/src/XMLValidator.cpp
#include "XMLValidator.hpp"

ValidChildren* XMLValidator::addSchemaElement(std::string element) {
    std::vector<ValidChildren*>::iterator schemaElementIterator =
        findSchemaElement(element);

    if (schemaElementIterator != schema.end()) {
        schema.erase(schemaElementIterator);
        delete *schemaElementIterator;
    }

    ValidChildren* schemaElement = 0;
    schema.push_back(schemaElement = new ValidChildren(element));
    return schemaElement;
}

std::vector<ValidChildren*>::iterator
XMLValidator::findSchemaElement(std::string element) {
    for (std::vector<ValidChildren*>::iterator i = schema.begin();
         i != schema.end(); i++) {
        if (((*i)->getThisElement().size() == 0 && element.size() == 0) ||
            ((*i)->getThisElement().size() > 0 &&
             (*i)->getThisElement().compare(element) == 0)) {
            return i;
        }
    }
    return schema.end();
}

void ValidChildren::addValidChild(const std::string& child, bool isAttribute) {
    if (childIsValid(child, isAttribute)) {
        return;
    }

    validChildren.push_back(child);
    childIsAttribute.push_back(isAttribute);
}

bool ValidChildren::childIsValid(const std::string& child, bool isAttribute) {
    for (unsigned int i = 0; i < validChildren.size(); i++) {
        if (childIsAttribute[i] == isAttribute &&
            validChildren[i].compare(child) == 0) {
            return true;
        }
    }
    return false;
}
